### DISCUSSION TITLE
How to implement netty channel handler to get request body

### DISCUSSION THREAD

**Discussion by ThraaxSession** [micronaut-projects/micronaut-core#8427] [Q&A]
Anyone knows how to implement `Netty channel handler` to get the request body for Micronaut 3.7.4?
See: https://github.com/micronaut-projects/micronaut-core/issues/7986

Currently the `HttpServerFilter` isn't able to supply the request body. As long the issue is open, I need to do it "manually".

@graemerocher as you mentioned the idea.

**Comment by yawkat** [comment]
the guide has an example of how to customize the channel pipeline: https://docs.micronaut.io/latest/guide/index.html#nettyServerPipeline

You can add your handler in `onRequestPipelineBuilt`, probably before `ChannelPipelineCustomizer.HANDLER_HTTP_STREAM`.

**Reply by ThraaxSession** [reply]
Hi @yawkat I don't get the docs actually. It looks like a partly explained example. Could you help me?
```kotlin
@Requires(beans = [SwsRouteFilter::class])
@Singleton
class SwsNettyServerCustomizer(
    private val swsRouteFilter: SwsRouteFilter
): BeanCreatedEventListener<NettyServerCustomizer.Registry> {
    override fun onCreated(event: BeanCreatedEvent<NettyServerCustomizer.Registry>): NettyServerCustomizer.Registry {
        val registry = event.bean
        registry.register(Customizer(null))
        return registry
    }

    private inner class Customizer constructor(private val channel: Channel?) : NettyServerCustomizer {
        override fun onStreamPipelineBuilt() {
            super.onStreamPipelineBuilt()
            channel!!.pipeline().addBefore(
                ChannelPipelineCustomizer.HANDLER_HTTP_STREAM,
                "SwsRouteFilter", SwsServerHandler(swsRouteFilter)
            )
        }
    }

    private inner class SwsServerHandler(
        private val swsRouteFilter: SwsRouteFilter
    ): ChannelDuplexHandler() {
        override fun channelRead(ctx: ChannelHandlerContext, msg: Any) {
            super.channelRead(ctx, msg)


        }
    }

}
````

**Reply by yawkat** [reply]
In SwsServerHandler you can intercept the various netty HttpObject subtypes, some of those will include the body. I think it is out of scope for the docs to explain how to write netty channel handlers.
